libpgcommon/lwgeom_transform.c

```diff
@@ -666,147 +797,144 @@ void SetPROJ4LibPath(void)
-#endif
 
-	if ( GetProjectionsUsingFCInfo(fcinfo, srid, srid, &pj1, &pj2) == LW_FAILURE)
-		return LW_FAILURE;
-
-	if ( ! pj_is_latlong(pj1) )
-		return LW_FAILURE;
-
-#if POSTGIS_PROJ_VERSION >= 48
-	/* For newer versions of Proj we can pull the spheroid paramaeters and initialize */
-	/* using them */
-	pj_get_spheroid_defn(pj1, &major_axis, &eccentricity_squared);
-	minor_axis = major_axis * sqrt(1-eccentricity_squared);
-	spheroid_init(s, major_axis, minor_axis);
+static int
+proj_pj_is_latlong(const PJ* pj)
+{
+#if POSTGIS_PROJ_VERSION < 60
+	return pj_is_latlong(pj->pj_from);
 #else
-	/* For old versions of Proj we cannot lookup the spheroid parameters from the API */
-	/* So we use the WGS84 parameters (boo!) */
-	spheroid_init(s, WGS84_MAJOR_AXIS, WGS84_MINOR_AXIS);
+	PJ_TYPE pj_type = proj_get_type(proj_get_source_crs(NULL, pj));
+	return (pj_type == PJ_TYPE_GEOGRAPHIC_2D_CRS) ||
+	       (pj_type == PJ_TYPE_GEOGRAPHIC_3D_CRS);
 #endif
-
-	return LW_SUCCESS;
 }
 
-void srid_is_latlong(FunctionCallInfo fcinfo, int srid)
+static int
+srid_is_latlong(FunctionCallInfo fcinfo, int srid)
 {
-	projPJ pj1;
-	projPJ pj2;
-
-	if ( srid == SRID_DEFAULT || srid == SRID_UNKNOWN )
-		return;
+	PJ* pj;
+	if ( GetPJUsingFCInfo(fcinfo, srid, srid, &pj) == LW_FAILURE)
+		return LW_FALSE;
+	return proj_pj_is_latlong(pj);
+}
 
-	if ( GetProjectionsUsingFCInfo(fcinfo, srid, srid, &pj1, &pj2) == LW_FAILURE)
+void
+srid_check_latlong(FunctionCallInfo fcinfo, int srid)
+{
+	if (srid == SRID_DEFAULT || srid == SRID_UNKNOWN)
 		return;
 
-	if ( pj_is_latlong(pj1) )
+	if (srid_is_latlong(fcinfo, srid))
 		return;
 
 	ereport(ERROR, (
 	            errcode(ERRCODE_INVALID_PARAMETER_VALUE),
 	            errmsg("Only lon/lat coordinate systems are supported in geography.")));
 }
 
-
-srs_precision srid_axis_precision(FunctionCallInfo fcinfo, int srid, int precision)
+srs_precision
+srid_axis_precision(FunctionCallInfo fcinfo, int srid, int precision)
 {
-	projPJ pj1;
-	projPJ pj2;
-
 	srs_precision sp;
-
 	sp.precision_xy = precision;
 	sp.precision_z = precision;
 	sp.precision_m = precision;
 
 	if ( srid == SRID_UNKNOWN )
 		return sp;
 
-	if ( GetProjectionsUsingFCInfo(fcinfo, srid, srid, &pj1, &pj2) == LW_FAILURE)
-		return sp;
-
-	if ( pj_is_latlong(pj1) )
+	if ( srid_is_latlong(fcinfo, srid) )
 	{
 		sp.precision_xy += 5;
 		return sp;
 	}
 
 	return sp;
 }
+
+int
+spheroid_init_from_srid(FunctionCallInfo fcinfo, int srid, SPHEROID *s)
+{
+	PJ* pj;
+#if POSTGIS_PROJ_VERSION >= 60
+	double out_semi_major_metre, out_semi_minor_metre, out_inv_flattening;
+	int out_is_semi_minor_computed;
+	PJ *pj_ellps, *pj_crs;
+#elif POSTGIS_PROJ_VERSION >= 48
+	double major_axis, minor_axis, eccentricity_squared;
+#endif
+
+	if ( GetPJUsingFCInfo(fcinfo, srid, srid, &pj) == LW_FAILURE)
+		return LW_FAILURE;
+
+#if POSTGIS_PROJ_VERSION >= 60
+	if (!proj_pj_is_latlong(pj))
+		return LW_FAILURE;
+	pj_crs = proj_get_source_crs(NULL, pj);
```

Comment:
`proj_get_source_crs` and `proj_get_ellipsoid` can return NULL, which would crash subsequent function calls.